### ISSUE TITLE
Node rendering needs more flexibility

### ISSUE THREAD

**Issue by runfaj** [cheton/react-infinite-tree#7]
Currently, you pass a renderer for each node. This would be great, but the tree cancels all click events, so you can't have buttons on a row, for simple things like add/edit/clone/delete on the node itself. It would be nice to rework the renderer to allow tapping into a parent component's state and allowing click listeners wherever you want them.

On this screenshot, the only button that works is the expand/collapse that utilizes the built-in tree expand/collapse. I can't tap into the parent's state at all to look at state variables, or click methods, so I can't hide and show those other buttons, nor actually click on them.

![image](https://user-images.githubusercontent.com/3292053/32796532-41d1aae0-c92c-11e7-8871-23d4b38c4f24.png)

Here's a very simple example of what I might expect for one of the buttons in jsx.

    {this.state.editing &&
        <div
            className={classNames('icon-button','delete')}
            onClick={this.deleteItem.bind(this, node)}
            title="Delete"
        >
            <i className="material-icons">close</i>
        </div>
    }

Is there any way this could get updated for the React project?

**Comment by cheton** [comment]
I made a quick PoC by hacking the tree's update method. Here is the original one:
https://github.com/cheton/infinite-tree/blob/master/src/infinite-tree.js#L1561-L1570
```js
update() {
    // Emit a "contentWillUpdate" event
    this.emit('contentWillUpdate');

    // Update list
    const rows = this.rows.filter(row => !!row);
    this.clusterize.update(rows);

    // Emit a "contentWillUpdate" event
    this.emit('contentDidUpdate');
}
```

This hack will hide the scrollElement and replace the update method. Each time `update` is called, it will call setState to update the nodes and trigger an extra rendering:

```js
this.tree = new InfiniteTree(options);
this.tree.scrollElement.style = 'display: none';

this.tree.update = () => {
    this.setState(state => ({
        nodes: this.tree.nodes
    });
};
```

In the render method, nodes will be rendered within the virtual list:
```js
import React from 'react';
import ReactDOM from 'react-dom';
import InfiniteTree from 'infinite-tree';
import VirtualList from 'react-tiny-virtual-list';
import { defaultRowRenderer } from './renderer';

const lcfirst = (str) => {
    str += '';
    return str.charAt(0).toLowerCase() + str.substr(1);
};

module.exports = class extends React.Component {
    tree = null;
    state = {
        nodes: []
    };

    eventHandlers = {
        onClick: null,
        onDoubleClick: null,
        onKeyDown: null,
        onKeyUp: null,
        onClusterWillChange: null,
        onClusterDidChange: null,
        onContentWillUpdate: null,
        onContentDidUpdate: null,
        onOpenNode: null,
        onCloseNode: null,
        onSelectNode: null,
        onWillOpenNode: null,
        onWillCloseNode: null,
        onWillSelectNode: null
    };

    componentDidMount() {
        const { children, className, style, ...options } = this.props;

        const el = ReactDOM.findDOMNode(this);
        options.el = el;

        options.rowRenderer = () => { return ''; };

        this.tree = new InfiniteTree(options);

        this.tree.scrollElement.style = 'display: none';

        this.tree.update = () => {
            this.setState(state => ({
                nodes: this.tree.nodes
            });
        };

        Object.keys(this.eventHandlers).forEach(key => {
            if (!this.props[key]) {
                return;
            }

            const eventName = lcfirst(key.substr(2)); // e.g. onContentWillUpdate -> contentWillUpdate
            this.eventHandlers[key] = this.props[key];
            this.tree.on(eventName, this.eventHandlers[key]);
        });
    }
    componentWillUnmount() {
        Object.keys(this.eventHandlers).forEach(key => {
            if (!this.eventHandlers[key]) {
                return;
            }

            const eventName = lcfirst(key.substr(2)); // e.g. onUpdate -> update
            this.tree.removeListener(eventName, this.eventHandlers[key]);
            this.eventHandlers[key] = null;
        });

        this.tree.destroy();
        this.tree = null;
    }
    render() {
        const { children, className, style } = this.props;
        const itemCount = this.tree ? this.tree.nodes.length : 0;

        return (
            <div className={className} style={style}>
                <VirtualList
                    ref={node => {
                        this.virtualList = node;
                    }}
                    width="100%"
                    height={400}
                    itemCount={itemCount}
                    itemSize={22}
                    renderItem={({ index, style }) => {
                        const node = this.tree.nodes[index];
                        const { rowRenderer } = this.props;
                        const row = rowRenderer(node, this.tree, this.tree.options);

                        return (
                            <div key={index} style={style}>
                                {row}
                            </div>
                        );
                    }}
                />
            </div>
        );
    }
};
```

Then I can bind onClick events and render Dropdown within the renderer:
![image](https://user-images.githubusercontent.com/447801/32829740-90088b52-c9b8-11e7-9233-85c0fc60a156.png)

**Comment by cheton** [comment]
I will try to divide infinite-tree into two parts, infinite-tree and infinite-tree-renderer, just like the way of using React and ReactDOM. Once I have cleaned up the code, it will easier to use any virtual scroll list solutions like clusterize.js, react-tiny-virtual-list, etc.

**Comment by runfaj** [comment]
Awesome - I think this makes sense going forward and could be very beneficial for creating a well-rounded out library. Thanks!!

**Comment by cheton** [comment]
I'm reworking the react-infinite-tree component, now you can handle events with React.

**Demo**
http://cheton.github.io/react-infinite-tree/

**Example**
https://github.com/cheton/react-infinite-tree/blob/master/examples/Tree.jsx
https://github.com/cheton/react-infinite-tree/blob/master/examples/render.jsx

```jsx
<InfiniteTree
    ref={node => {
        this.tree = node ? node.tree : null;
    }}
    style={{
        border: '1px solid #ccc'
    }}
    data={this.data}
    width="100%"
    height={400}
    rowHeight={30}
    autoOpen={true}
    loadNodes={(parentNode, done) => {
        const suffix = parentNode.id.replace(/(\w)+/, '');
        const nodes = [
            {
                id: 'node1' + suffix,
                name: 'Node 1'
            },
            {
                id: 'node2' + suffix,
                name: 'Node 2'
            }
        ];
        setTimeout(() => {
            done(null, nodes);
        }, 1000);
    }}
    selectable={true} // Defaults to true
    shouldSelectNode={(node) => { // Defaults to null
        if (!node || (node === this.tree.getSelectedNode())) {
            return false; // Prevent from deselecting the current node
        }
        return true;
    }}
    onKeyDown={(event) => {
        const target = event.target || event.srcElement; // IE8
        console.log('onKeyDown', target);
        event.preventDefault();

        const node = this.tree.getSelectedNode();
        const nodeIndex = this.tree.getSelectedIndex();

        if (event.keyCode === 37) { // Left
            this.tree.closeNode(node);
        } else if (event.keyCode === 38) { // Up
            const prevNode = this.tree.nodes[nodeIndex - 1] || node;
            this.tree.selectNode(prevNode);
        } else if (event.keyCode === 39) { // Right
            this.tree.openNode(node);
        } else if (event.keyCode === 40) { // Down
            const nextNode = this.tree.nodes[nodeIndex + 1] || node;
            this.tree.selectNode(nextNode);
        }
    }}
>
{({ node, tree }) => {
    const { id, name, loadOnDemand = false, children, state, props = {} } = node;
    const { depth, open, path, total, loading = false, selected = false } = state;
    const childrenLength = Object.keys(children).length;
    const more = node.hasChildren();

    let togglerState = '';
    if ((!more && loadOnDemand) || (more && !open)) {
        togglerState = 'collapsed';
    }
    if (more && open) {
        togglerState = 'expanded';
    }

    let iconState = '';
    if (more && open) {
        iconState = 'folder-open';
    }
    if (more && !open) {
        iconState = 'folder';
    }
    if (!more) {
        iconState = 'file';
    }

    return (
        <TreeNode
            selected={selected}
            depth={depth}
            onClick={event => {
                tree.selectNode(node);
            }}
        >
            <Toggler
                state={togglerState}
                onClick={() => {
                    if (togglerState === 'collapsed') {
                        tree.openNode(node);
                    } else if (togglerState === 'expanded') {
                        tree.closeNode(node);
                    }
                }}
            />
            <PointerCursor>
                <Icon state={iconState} />
                <Text>{name}</Text>
            </PointerCursor>
        </TreeNode>
    );
}}
</InfiniteTree>
```

**Comment by cheton** [comment]
I will update the documentation and publish v1.0 release in the next few days.

**Comment by runfaj** [comment]
Thanks for getting something put together so quickly - you're awesome!

**Comment by cheton** [comment]
Released v1.0.0-alpha. You can see README for the usage.

Demo: http://cheton.github.io/react-infinite-tree/
